klippy/webhooks.py

```diff
@@ -0,0 +1,245 @@
+        self.is_server_connected = True
+        self.fd_handler = self.reactor.register_fd(
+            self.socket.fileno(), self.process_received)
+        self.mutex = self.reactor.mutex()
+        printer.register_event_handler('klippy:disconnect', self.close_socket)
+
+    def close_socket(self):
+        if self.is_server_connected:
+            logging.info("ServerConnection: lost connection to Moonraker")
+            self.is_server_connected = False
+            self.reactor.unregister_fd(self.fd_handler)
+            try:
+                self.socket.close()
+            except socket.error:
+                pass
+
+    def is_connected(self):
+        return self.is_server_connected
+
+    def process_received(self, eventtime):
+        try:
+            data = self.socket.recv(4096)
+        except socket.error as e:
+            # If bad file descriptor allow connection to be
+            # closed by the data check
+            if e.errno == 9:
+                data = ''
+            else:
+                return
+        if data == '':
+            # Socket Closed
+            self.close_socket()
+            return
+        requests = data.split('\x00')
+        requests[0] = self.partial_data + requests[0]
+        self.partial_data = requests.pop()
+        for req in requests:
+            logging.debug(
+                "ServerConnection: Request received from Moonraker %s" % (req))
+            try:
+                decoded_req = json_loads_byteified(req)
+                self._process_request(decoded_req)
+            except Exception:
+                logging.exception(
+                    "ServerConnection: Error processing Server Request %s"
+                    % (req))
+
+    def _process_request(self, req):
+        web_request = WebRequest(req)
+        try:
+            func = self.webhooks.get_callback(
+                web_request.get_path())
+            func(web_request)
+        except WebRequestError as e:
+            web_request.set_error(e)
+        except Exception as e:
+            web_request.set_error(WebRequestError(e.message))
+        result = web_request.finish()
+        logging.debug(
+            "ServerConnection: Sending response - %s" % (str(result)))
+        self.send({'method': "response", 'params': result})
+
+    def send(self, data):
+        if not self.is_server_connected:
+            return
+        with self.mutex:
+            retries = 10
+            data = json.dumps(data) + "\x00"
+            while data:
+                try:
+                    sent = self.socket.send(data)
+                except socket.error as e:
+                    if e.errno == 9 or e.errno == 32 or not retries:
+                        sent = 0
+                    else:
+                        retries -= 1
+                        waketime = self.reactor.monotonic() + .001
+                        self.reactor.pause(waketime)
```

Comment:
FYI, if send() can block it could make the code handlers harder to write (as they then have to be aware that they may get rescheduled and possibly get called reentrant).  Not something to fix now, but something to think about after merging.

Reply:
The mutex will protect the send() method from being called reentrant (so responses wont be interleaved).  However, it doesn't prevent the caller from being invoked reentrant.  So, for example, lets say someone made a `my_module/my_action` endpoint that did something like: ` do_some_prep_work(); send("starting work"); do_more_work()` - then there is the possibility that send() blocks and `my_module/my_action` gets called again - leading to a code path like: `do_some_prep_work() ; send_start... ; do_some_prep_work() ; send() do_more_work(); send_complete... ; do_more_work()`.

Reply:
Ah, I see.  I'll think about this going forward.